File: jdk13/src/io/github/karlatemp/jll/j13/OverrideReflectionFactory.java
/*
 * Copyright (c) 2018-2020 Karlatemp. All rights reserved.
 * @author Karlatemp <karlatemp@vip.qq.com> <https://github.com/Karlatemp>
 * @create 2020/05/17 13:48:43
 *
 * JVMLowLevel/jdk13/OverrideReflectionFactory.java
 */

package io.github.karlatemp.jll.j13;

import jdk.internal.misc.Unsafe;
import jdk.internal.reflect.ConstructorAccessor;
import jdk.internal.reflect.FieldAccessor;
import jdk.internal.reflect.MethodAccessor;
import jdk.internal.reflect.ReflectionFactory;

import java.lang.invoke.MethodHandle;
import java.lang.invoke.MethodHandles;
import java.lang.reflect.Constructor;
import java.lang.reflect.Field;
import java.lang.reflect.Method;
import java.util.HashSet;
import java.util.Set;

@SuppressWarnings("rawtypes")
public class OverrideReflectionFactory extends ReflectionModel {
    public static final Set<Class<?>> unsafe = new HashSet<>(Set.of(
            Class.class,
            ReflectionFactory.class,
            Unsafe.class,
            sun.misc.Unsafe.class,
            MethodHandles.class,
            MethodHandles.Lookup.class,
            MethodHandle.class
    ));

    public static boolean isUnsafe(Class<?> klass) {
        for (var c : unsafe) {
            if (c.isAssignableFrom(klass)) return true;
        }
        return false;
    }

    @Override
    public ConstructorAccessor newConstructorAccessor(Constructor var0) {
        if (isUnsafe(var0.getDeclaringClass()))
            return DenyAccessorImpl.instance;
        return super.newConstructorAccessor(var0);
    }

    @Override
    public MethodAccessor newMethodAccessor(Method var0) {
        if (isUnsafe(var0.getReturnType()) || isUnsafe(var0.getDeclaringClass()))
            return DenyAccessorImpl.instance;
        return super.newMethodAccessor(var0);
    }

    @Override
    public FieldAccessor newFieldAccessor(Field var0, boolean var1) {
        if (isUnsafe(var0.getType()) || isUnsafe(var0.getDeclaringClass()))
            return DenyAccessorImpl.instance;
        return super.newFieldAccessor(var0, var1);
    }
}

File: src/io/github/karlatemp/jll/OverrideReflectionFactory.kt
/*
 * Copyright (c) 2018-2020 Karlatemp. All rights reserved.
 * @author Karlatemp <karlatemp@vip.qq.com> <https://github.com/Karlatemp>
 * @create 2020/05/17 13:48:43
 *
 * JVMLowLevel/JVMLowLevel/OverrideReflectionFactory.kt
 */

package io.github.karlatemp.jll

import sun.misc.Unsafe
import sun.reflect.ConstructorAccessor
import sun.reflect.FieldAccessor
import sun.reflect.MethodAccessor
import sun.reflect.ReflectionFactory
import java.lang.invoke.MethodHandle
import java.lang.invoke.MethodHandles
import java.lang.reflect.Constructor
import java.lang.reflect.Field
import java.lang.reflect.Method

// Don't use object, it is not support
class OverrideReflectionFactory private constructor() : ReflectionFactoryModel() {
    private fun isUnsafe(klass: Class<*>): Boolean {
        UnsafeAccessConst.unsafe.forEach {
            if (it.isAssignableFrom(klass)) return@isUnsafe true
        }
        return false
    }

    override fun newMethodAccessor(var0: Method): MethodAccessor {
        if (isUnsafe(var0.declaringClass) || isUnsafe(var0.returnType))
            return DenyAccessor
        if (var0.declaringClass?.`package`?.name == "io.github.karlatemp.jll") {
            return DenyAccessor
        }
        return super.newMethodAccessor(var0)
    }

    override fun newConstructorAccessor(var0: Constructor<*>): ConstructorAccessor {
        if (isUnsafe(var0.declaringClass))
            return DenyAccessor
        if (var0.declaringClass == MethodHandles.Lookup::class.java) {
            return DenyAccessor
        }
        return super.newConstructorAccessor(var0)
    }

    override fun newFieldAccessor(var0: Field, var1: Boolean): FieldAccessor {
        if (isUnsafe(var0.declaringClass) || isUnsafe(var0.type))
            return DenyAccessor
        return super.newFieldAccessor(var0, var1)
    }
}